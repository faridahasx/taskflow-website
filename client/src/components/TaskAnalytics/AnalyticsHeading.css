#analytics-heading {
  width: 100%;
  padding-top: 8px;
  align-items: center;
  justify-content: space-between;
}

#analytics-dt-container {
  font-weight: 500;
  font-size: 15px;
  font-style: italic;
}

.datetime {
  margin: 0 5px;
}

@media all and (max-width: 380px) {
  #analytics-heading {
    align-items: start;
    flex-direction: column;
  }

  .datetime {
    margin: 5px 5px 5px 0;
  }

  .datetime:nth-child(2) {
    margin-right: 5px;
  }
}
