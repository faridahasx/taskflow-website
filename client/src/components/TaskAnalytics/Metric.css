.metric-li {
  height: 40px;
  padding: 0 20px;
  white-space: nowrap;
  align-items: center;
}

.metric-key {
  width: 50%;
  min-width: 200px;
  font-weight: 500;
}

.metric-value {
  width: 50%;
}

.metric-li:first-child,
.metric-li:first-child .metric-key,
.metric-li:last-child,
.metric-li:last-child .metric-key {
  font-weight: 600;
}

@media all and (min-width: 400px) {
  .metric-li {
    border-bottom: 1px solid rgba(128, 128, 128, 0.363);
  }

  .metric-li:last-child {
    border-top: 1px solid rgb(128, 128, 128);
  }

  .metric-li:first-child {
    border-bottom-color: rgb(128, 128, 128);
  }
}
