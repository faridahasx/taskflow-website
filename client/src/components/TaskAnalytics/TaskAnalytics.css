#analytics {
  width: 100%;
  min-height: calc(var(--vh) - 170px);
  margin-top: 40px;
  padding: 40px 0;
  justify-content: center;
  align-items: center;
}

#analytics .icon {
  border-color: rgba(0, 0, 0, 0);
}

#analytics-wrapper {
  width: calc(100vw - 40px);
  max-width: 800px;
  min-width: calc(var(--min-width) - 20px);
  border-radius: 10px;
  background-color: rgba(255, 255, 255, 0.922);
  box-shadow:
    0 4px 8px -10px rgba(100, 94, 94, 0.604),
    0 6px 20px 0 rgba(0, 0, 0, 0.275);
}

.analytics-empty {
  height: 280px;
}

#analytics-top {
  align-items: start;
  padding: 5px 10px;
  width: 100%;
}

#analytics-top h1 {
  color: var(--main-color-dark);
  font-size: 19px;
}

#analytics-summary {
  width: 100%;
  overflow: auto;
  border-top: 1px solid rgb(128, 128, 128);
}

#analytics-summary::-webkit-scrollbar-track {
  margin: 0 10px;
}

@media all and (min-width: 900px) {
  #analytics-wrapper {
    margin-left: 50px;
  }
}
