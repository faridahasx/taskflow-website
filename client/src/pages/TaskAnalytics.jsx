import TaskAnalyticsContainer from "containers/TaskAnalyticsContainer";
import Layout from "components/Layout/Layout";

const TaskAnalytics = () => {
  return (
    <Layout>
      <TaskAnalyticsContainer />
    </Layout>
  );
};

export default TaskAnalytics;
