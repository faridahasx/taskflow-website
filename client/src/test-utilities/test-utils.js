import { render } from "@testing-library/react";
import { TasksDispatchContext } from "context/TaskContext";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";

const customRender = (ui, { props, ...options } = {}) => {
  const { store, mockDispatchTasks } = props;

  const AllTheProviders = ({ children }) => (
    <Provider store={store}>
      <TasksDispatchContext.Provider value={mockDispatchTasks}>
        <BrowserRouter>{children}</BrowserRouter>
      </TasksDispatchContext.Provider>
    </Provider>
  );

  return render(ui, { wrapper: AllTheProviders, ...options });
};

export * from "@testing-library/react";
// export * from "@testing-library/jest-dom";
export { customRender as render };
