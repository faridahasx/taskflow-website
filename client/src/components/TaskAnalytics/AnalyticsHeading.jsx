import "./AnalyticsHeading.css";

const AnalyticsHeading = ({ start, end }) => {
  return (
    <h1 id="analytics-heading" className="flex">
      Task Analytics:
      {start && (
        <span id="analytics-dt-container" className="center">
          <span className="center datetime">
            <time dateTime={start}>{start}</time>
          </span>{" "}
          -
          <span className="center datetime">
            <time dateTime={end}>{end}</time>
          </span>
        </span>
      )}
    </h1>
  );
};

export default AnalyticsHeading;
