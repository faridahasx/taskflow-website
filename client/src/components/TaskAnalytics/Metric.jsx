// External imports
import PropTypes from "prop-types";
// Styles
import "./Metric.css";

const Metric = ({ metricKey, metricValue }) => {
  return (
    <li className="metric-li flex">
      <span className="metric-key flex">{metricKey}:</span>
      <span className="metric-value flex">{metricValue}</span>
    </li>
  );
};

Metric.propTypes = {
  metricKey: PropTypes.string.isRequired,
  metricValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
    .isRequired,
};
export default Metric;
