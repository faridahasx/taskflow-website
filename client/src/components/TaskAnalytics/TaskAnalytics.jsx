// External imports
import { useMemo } from "react";
import PropTypes from "prop-types";
// Hooks
import useNetworkStatus from "hooks/useNetworkStatus";
// Components
import OfflineContent from "components/OfflineContent";
import CircularLoading from "components/Loading/CircularLoading";
import FiltersNav from "components/NavigationItem/FiltersNav";
import TryAgain from "components/IconButtons/TryAgain";
import Metric from "./Metric";
import AnalyticsHeading from "./AnalyticsHeading";
// Styles
import "./TaskAnalytics.css";

const TaskAnalytics = ({
  start,
  end,
  taskAnalytics,
  errorDuringFetch,
  handleTryAgain,
}) => {
  const metricNames = useMemo(
    () => taskAnalytics && Object.keys(taskAnalytics),
    [taskAnalytics]
  );
  const isOnline = useNetworkStatus();

  return (
    <div
      role="region"
      aria-labelledby="analytics-heading"
      id="analytics"
      className="column"
    >
      {!isOnline && taskAnalytics === null ? (
        <OfflineContent />
      ) : (
        <div id="analytics-wrapper" className="column">
          <>
            <div id="analytics-top" className="flex">
              <FiltersNav />
              <AnalyticsHeading start={start} end={end} />
            </div>

            {taskAnalytics === null ? (
              <span className="center analytics-empty">
                {errorDuringFetch ? (
                  <TryAgain onClick={handleTryAgain} />
                ) : (
                  <CircularLoading />
                )}
              </span>
            ) : (
              taskAnalytics !== null &&
              (metricNames.length ? (
                <ul
                  aria-label="Task Analytics Summary"
                  id="analytics-summary"
                  className=""
                >
                  {metricNames.map((s) => (
                    <Metric
                      key={s}
                      metricKey={s}
                      metricValue={taskAnalytics[s]}
                    />
                  ))}
                </ul>
              ) : (
                <span className="center analytics-empty">
                  No available data
                </span>
              ))
            )}
          </>
        </div>
      )}
    </div>
  );
};

TaskAnalytics.propTypes = {
  start: PropTypes.string.isRequired,
  end: PropTypes.string.isRequired,
  taskAnalytics: PropTypes.any,
  errorDuringFetch: PropTypes.bool.isRequired,
  handleTryAgain: PropTypes.func.isRequired,
};

export default TaskAnalytics;
