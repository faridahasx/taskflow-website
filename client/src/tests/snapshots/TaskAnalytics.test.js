import { render } from "test-utilities/test-utils";
import mockInitialState, {
  mockStore,
} from "test-utilities/mocks/mockReduxState";
import TaskAnalytics from "pages/TaskAnalytics";

test("TaskAnalytics snapshot", () => {
  const { asFragment } = render(<TaskAnalytics />, {
    props: { store: mockStore(mockInitialState) },
  });
  expect(asFragment()).toMatchSnapshot("TaskAnalytics");
});
