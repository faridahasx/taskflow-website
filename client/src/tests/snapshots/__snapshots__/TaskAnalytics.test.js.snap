// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskAnalytics snapshot: TaskAnalytics 1`] = `
<DocumentFragment>
  <header
    class="flex"
    id="header"
  >
    <span
      class="flex"
      id="header-first"
    >
      <button
        aria-label="Open Categories"
        class="icon "
        title="Open Categories"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="MenuIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
          />
        </svg>
      </button>
      <a
        class="center"
        data-testid="header-home-link"
        href="/"
      >
        <div
          aria-label="TaskFlow Logo"
          class="logo center"
          role="img"
          title="TaskFlow"
        >
          <span
            class="logo-task"
          >
            Task
          </span>
          <span
            class="logo-flow"
          >
            Flow
          </span>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="MovingIcon"
            focusable="false"
            id="logo-icon"
            viewBox="0 0 24 24"
          >
            <path
              d="M19.71 9.71 22 12V6h-6l2.29 2.29-4.17 4.17c-.39.39-1.02.39-1.41 0l-1.17-1.17c-1.17-1.17-3.07-1.17-4.24 0L2 16.59 3.41 18l5.29-5.29c.39-.39 1.02-.39 1.41 0l1.17 1.17c1.17 1.17 3.07 1.17 4.24 0l4.19-4.17z"
            />
          </svg>
        </div>
      </a>
    </span>
    <span
      class="flex"
      id="header-second"
    >
      <a
        class="icon"
        data-testid="header-analytics-link"
        href="/analytics"
        title="Task Analytics"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="BarChartIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M4 9h4v11H4zm12 4h4v7h-4zm-6-9h4v16h-4z"
          />
        </svg>
      </a>
      <button
        class="icon "
        title="Sign Out"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="LogoutTwoToneIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M5 5h7V3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h7v-2H5V5zm16 7-4-4v3H9v2h8v3l4-4z"
          />
        </svg>
      </button>
    </span>
  </header>
  <main>
    <div
      class="categories-slider "
      data-testid="categories"
      id="categories-dropdown"
    >
      <div
        class="categories "
      >
        <div
          class="center"
          id="categories-list-container"
        >
          <span
            class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
            data-testid="loading"
            role="progressbar"
            style="width: 40px; height: 40px;"
          >
            <svg
              class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
              viewBox="22 22 44 44"
            >
              <circle
                class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
                cx="44"
                cy="44"
                fill="none"
                r="20.2"
                stroke-width="3.6"
              />
            </svg>
          </span>
        </div>
        <button
          aria-label="Add new category"
          class="center"
          id="add-category"
          title="Add new category"
        >
          <span
            class="center"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="AddIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
              />
            </svg>
          </span>
          New Category
        </button>
      </div>
    </div>
    <div
      aria-labelledby="analytics-heading"
      class="column"
      id="analytics"
      role="region"
    >
      <div
        class="column"
        id="analytics-wrapper"
      >
        <div
          class="flex"
          id="analytics-top"
        >
          <span
            class="nav-item center"
          >
            <button
              class="icon "
              title="Filters"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="FilterListIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M10 18h4v-2h-4v2zM3 6v2h18V6H3zm3 7h12v-2H6v2z"
                />
              </svg>
            </button>
          </span>
          <h1
            class="flex"
            id="analytics-heading"
          >
            Task Analytics:
          </h1>
        </div>
        <span
          class="center analytics-empty"
        >
          <span
            class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
            data-testid="loading"
            role="progressbar"
            style="width: 40px; height: 40px;"
          >
            <svg
              class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
              viewBox="22 22 44 44"
            >
              <circle
                class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
                cx="44"
                cy="44"
                fill="none"
                r="20.2"
                stroke-width="3.6"
              />
            </svg>
          </span>
        </span>
      </div>
    </div>
  </main>
  <footer
    class="center"
  >
    <div
      class="center"
      id="footer-div"
    >
      <a
        aria-label="Go to home page"
        href="/"
        id="footer-logo-link"
        title="Home"
      >
        <div
          aria-label="TaskFlow Logo"
          class="logo center"
          role="img"
          title="TaskFlow"
        >
          <span
            class="logo-task"
          >
            Task
          </span>
          <span
            class="logo-flow"
          >
            Flow
          </span>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="MovingIcon"
            focusable="false"
            id="logo-icon"
            viewBox="0 0 24 24"
          >
            <path
              d="M19.71 9.71 22 12V6h-6l2.29 2.29-4.17 4.17c-.39.39-1.02.39-1.41 0l-1.17-1.17c-1.17-1.17-3.07-1.17-4.24 0L2 16.59 3.41 18l5.29-5.29c.39-.39 1.02-.39 1.41 0l1.17 1.17c1.17 1.17 3.07 1.17 4.24 0l4.19-4.17z"
            />
          </svg>
        </div>
      </a>
      <a
        href="https://farida-mu.vercel.app/#contact"
        id="dev-link"
        target="_blank"
        title="Contact the developer"
      >
        Contact the developer
      </a>
    </div>
  </footer>
</DocumentFragment>
`;
